[#48717] Replace DelayedJob with GoodJob.

ba1ash opened the pull request:
https://community.openproject.org/work_packages/48717

**TODOS**:

- [x] Create PR for SaaS changes. It extends ApplicationJob (delayed_job_extender.rb) with tenant specific code and needs to be adapted
- [x] adapt https://github.com/opf/openproject/blob/dev/docker/prod/worker and https://github.com/opf/openproject/blob/dev/docker/prod/bimworker  to work with good job
- [x] adapt https://github.com/opf/openproject/blob/dev/Procfile and https://github.com/opf/openproject/blob/dev/Procfile.dev
- [x] Adapt  documentation that uses `rake jobs:work`
- [x] Adapt health_checks

ba1ash committed:
[#48717] Replace DelayedJob with GoodJob.

https://community.openproject.org/work_packages/48717

oliverguenther commented:
Hi @ba1ash, thanks for you efforts. I've added a number of todos that need to happen both in the core and the saas-openproject repo to be able to get this merged.

ba1ash committed:
Merge branch 'dev' into task/48717-replace-delayedjob

ba1ash committed:
Modify good_job_settings only if required.

apfohl commented on app/contracts/settings/working_days_params_contract.rb:
```diff
@@ -41,8 +41,11 @@ def working_days_are_present
       end
     end
 
+    # TODO: consider implementing using GoodJob concurrency control mechanisms
```
Can you create a work package for this?

ba1ash replied:
Sure, done. https://community.openproject.org/wp/53122

apfohl commented on config/application.rb:
```diff
@@ -212,7 +212,18 @@ class Application < Rails::Application
+
+    config.good_job.preserve_job_records = true
+    config.good_job.retry_on_unhandled_error = false
+    # config.good_job.on_thread_error = -> (exception) { Rails.error.report(exception) }
```
Why not log it?

machisuji replied:
Agreed. It probably makes sense to pass this to the OpenProject logger which then can report it to appsignal.

ba1ash replied:
It seems we do not this option to get notification in Appsignal. It is already done by [Appsginal ActiveJob extension](https://github.com/appsignal/appsignal-ruby/blob/4f907dc48cb77defebe06d65a0089a1aed7d2468/lib/appsignal/hooks/active_job.rb#L57): 
And ActiveJob::Base has it included: 
```bash
[7] pry(main)> ActiveJob::Base.singleton_class.ancestors
=> [#<Class:ActiveJob::Base>,
 Appsignal::Hooks::ActiveJobHook::ActiveJobClassInstrumentation,
.....
 Kernel,
 BasicObject]
[8] pry(main)>
```
If I enable it then error reports are duplicated in Appsignal.

apfohl commented on docs/development/development-environment-docker/README.md:
```diff
@@ -141,15 +141,8 @@ system's resources.
 docker compose up -d worker
-
-# Start the worker service to work off all delayed jobs and shut it down afterwards
-docker compose run --rm worker rake jobs:workoff
```
Is there a similar thing for Good Job?

ba1ash replied:
I have not found an analogy for that :(

oliverguenther commented on config/routes.rb:
```diff
@@ -639,4 +639,8 @@
     mount Lookbook::Engine, at: "/lookbook"
   end
+
+  constraints(->(req) { User.exists?(id: req.session[:user_id], admin: true) }) do
```
This needs to be configurable. On SaaS, you'd otherwise see all jobs in all schemas. Also, the engine allows you to inspect all attributes of a job. In some cases, this might include sensitive arguments. So I'd vote to for now only enable this on development mode.

ba1ash replied:
True. With current modification to SaaS the dashboard will have some bugs there. So, I left it enabled for dev only.

oliverguenther commented on db/migrate/20220202140507_reorder_project_children.rb:
```diff
@@ -1,33 +0,0 @@
-
-class ReorderProjectChildren < ActiveRecord::Migration[6.1]
-  def up
-    ::Projects::ReorderHierarchyJob.perform_later
```
Why did you remove this migration? This would still be necessary on old installations. I can see the point about removing delayed job cron removals, but this should remain

oliverguenther replied:
The migration now fails as good job is not yet defined. You can just run the job in the migration without going through good_job

ba1ash replied:
Good point!
I wonder if that is okay to have it asynchronous here. Will the job be executed eventually if this migration happens alongside with GoodJob migration? 
I expect the code to be new when running this migration. It means that `::Projects::ReorderHierarchyJob.perform_later` will try to reach GoodJob tables which are absent at the moment, because good_job migrations are younger. If that is true there will be an error. I suggest to run the job synchronously to avoid the error then. Does it make sense to you?

By removing I tried to solve this problem which apparently is not a solution :)

ba1ash replied:
Let's rewrite it like that? I am not aware of how heavy the job can be...
```
class ReorderProjectChildren < ActiveRecord::Migration[6.1]
  def up
    ::Projects::ReorderHierarchyJob.new.perform
  end
end  
```

oliverguenther commented on db/migrate/20220804112533_remove_notification_cleanup_job.rb:
```diff
@@ -1,41 +0,0 @@
-      .where('handler LIKE ?', "%job_class: Notifications::CleanupJob%")
-      .delete_all
-
-    Setting
```
Same here, this part needs to be retained

machisuji replied:
This is not resolved, is it? The code is still removed. But of course Delayed::Job won't exist anymore. So this should just be done in plain SQL.

```
execute_sql "DELETE FROM delayed_jobs WHERE handler LIKE '%job_class: Notifications::CleanupJob%'"
```

oliverguenther commented:
- [ ] Trying to instantiate a project copy, I got this error on first try: 

```
E, [2024-02-19T20:37:15.941960 #83213] ERROR -- : [ActiveJob] [CopyProjectJob] [0e0cdbb3-af42-4717-b907-a05249a89ef6] Error performing CopyProjectJob (Job ID: 0e0cdbb3-af42-4717-b907-a05249a89ef6) from GoodJob(default) in 54.13ms: ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_delayed_job_statuses_on_job_id"
DETAIL:  Key (job_id)=(0e0cdbb3-af42-4717-b907-a05249a89ef6) already exists.
):
```

oliverguenther committed:
Replace background check with good_job check

oliverguenther committed:
Extract health checks into separate files

ba1ash committed:
Enable GoodJob dashboard only in development.

ba1ash committed:
Resurrect removed migrations.

oliverguenther commented:
> * [ ]  Trying to instantiate a project copy, I got this error on first try:
> 
> ```
> E, [2024-02-19T20:37:15.941960 #83213] ERROR -- : [ActiveJob] [CopyProjectJob] [0e0cdbb3-af42-4717-b907-a05249a89ef6] Error performing CopyProjectJob (Job ID: 0e0cdbb3-af42-4717-b907-a05249a89ef6) from GoodJob(default) in 54.13ms: ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_delayed_job_statuses_on_job_id"
> DETAIL:  Key (job_id)=(0e0cdbb3-af42-4717-b907-a05249a89ef6) already exists.
> ):
> ```

I couldn't reproduce this anymore. It might be an error on my side due to the way I ran jobs manually

ba1ash commented:
> > * [ ]   Trying to instantiate a project copy, I got this error on first try:
> > 
> > ```
> > E, [2024-02-19T20:37:15.941960 #83213] ERROR -- : [ActiveJob] [CopyProjectJob] [0e0cdbb3-af42-4717-b907-a05249a89ef6] Error performing CopyProjectJob (Job ID: 0e0cdbb3-af42-4717-b907-a05249a89ef6) from GoodJob(default) in 54.13ms: ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_delayed_job_statuses_on_job_id"
> > DETAIL:  Key (job_id)=(0e0cdbb3-af42-4717-b907-a05249a89ef6) already exists.
> > ):
> > ```
> 
> I couldn't reproduce this anymore. It might be an error on my side due to the way I ran jobs manually

Probably, I tested locally and it works fine.

ba1ash committed:
Run ReorderHierarchyJob synchronously in migration

oliverguenther committed:
Remove the reorder_hierarchy_job and inline it in the migration

oliverguenther reviewed (approved):
@ba1ash great work, this is looking good to me in the core. SaaS/Deployment is missing some changes that I need to check with Markus tomorrow. So let's hold off merging this just yet.

ba1ash commented:
> @ba1ash great work, this is looking good to me in the core. SaaS/Deployment is missing some changes that I need to check with Markus tomorrow. So let's hold off merging this just yet.

Great! Thanks for reviewing @oliverguenther!

oliverguenther commented:
> * [ ]  Trying to instantiate a project copy, I got this error on first try:
> 
> ```
> E, [2024-02-19T20:37:15.941960 #83213] ERROR -- : [ActiveJob] [CopyProjectJob] [0e0cdbb3-af42-4717-b907-a05249a89ef6] Error performing CopyProjectJob (Job ID: 0e0cdbb3-af42-4717-b907-a05249a89ef6) from GoodJob(default) in 54.13ms: ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_delayed_job_statuses_on_job_id"
> DETAIL:  Key (job_id)=(0e0cdbb3-af42-4717-b907-a05249a89ef6) already exists.
> ):
> ```

@ba1ash  This error happened to me again just now, and to Markus as well. I think there is something wrong here after all.

machisuji commented:
It looks good over all (also the multitenancy PR) but I'm still looking into why I'm getting duplicate key errors on upserting the job status.

On a positive side note, I've tested the case where the worker is killed while working off a job.
In delayed job that job will just be lost, because there is only the "locked_by" field. And when the worker is restarted the job won't be re-tried again because the job is already locked.

Good job has two separate columns, started_at and finished_at so it when the worker is killed, the latter one will not be set and the job will be re-retried eventually. That's pretty neat, just like the job dashboard. Really nice!

![image](https://github.com/opf/openproject/assets/158871/75bbb08c-a8c8-4643-8495-30248216d106)

ba1ash commented:
> > * [ ]   Trying to instantiate a project copy, I got this error on first try:
> > 
> > ```
> > E, [2024-02-19T20:37:15.941960 #83213] ERROR -- : [ActiveJob] [CopyProjectJob] [0e0cdbb3-af42-4717-b907-a05249a89ef6] Error performing CopyProjectJob (Job ID: 0e0cdbb3-af42-4717-b907-a05249a89ef6) from GoodJob(default) in 54.13ms: ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_delayed_job_statuses_on_job_id"
> > DETAIL:  Key (job_id)=(0e0cdbb3-af42-4717-b907-a05249a89ef6) already exists.
> > ):
> > ```
> 
> @ba1ash This error happened to me again just now, and to Markus as well. I think there is something wrong here after all.

Okay, I will try to walk through the copying process more carefully.

machisuji commented:
> > > * [ ]    Trying to instantiate a project copy, I got this error on first try:
> > > 
> > > ```
> > > E, [2024-02-19T20:37:15.941960 #83213] ERROR -- : [ActiveJob] [CopyProjectJob] [0e0cdbb3-af42-4717-b907-a05249a89ef6] Error performing CopyProjectJob (Job ID: 0e0cdbb3-af42-4717-b907-a05249a89ef6) from GoodJob(default) in 54.13ms: ActiveRecord::RecordNotUnique (PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_delayed_job_statuses_on_job_id"
> > > DETAIL:  Key (job_id)=(0e0cdbb3-af42-4717-b907-a05249a89ef6) already exists.
> > > ):
> > > ```
> > 
> > 
> > @ba1ash This error happened to me again just now, and to Markus as well. I think there is something wrong here after all.
> 
> Okay, I will try to walk through the copying process more carefully.

I'm thinking this is a multithreading issue. I can reproduce the problem quite often.
The code fails in `ApplicationJobWithStatus#upsert_status` where it tries to update the job status to `in_process` [sic].
There while there already is a record with the `'in_queue`' status, `resource.new_record?` is actually true. Which is not correct.

We may need some sort of double-checked locking here.
We could simply catch `ActiveRecord::RecordNotUnique` [here](https://github.com/opf/openproject/blob/dev/modules/job_status/app/workers/job_status/application_job_with_status.rb#L80) and re-try the whole method. That would probably already solve the issue.

Although the better solution, IMHO, would be to use plain SQL and use what amounts to the following during the update statement.

```
ON CONFLICT (job_id) DO UPDATE SET serialized_params = ?
```

machisuji commented on app/contracts/settings/working_days_params_contract.rb:
```diff
@@ -41,8 +41,11 @@ def working_days_are_present
+    # TODO: consider implementing using GoodJob concurrency control mechanisms
     def unique_job
-      if WorkPackages::ApplyWorkingDaysChangeJob.scheduled?
+      if GoodJob::Job
```
I would suggest

1) putting this into a method like `job_scheduled?'
2) using `job_class: WorkPackages::ApplyWorkingDaysChangeJob.name`

The latter makes sure we don't miss this as easily if the job was ever to be renamed or removed because this will then lead to a runtime error rather than the query just silently returning nothing always.

machisuji commented on app/workers/application_job.rb:
```diff
@@ -92,6 +92,11 @@ def reload_mailer_settings!
     Setting.reload_mailer_settings!
   end
 
+  def good_job_scheduled_at
```
I would argue that we don't want that implementation detail of it being good job in the method name here. I'd prefer `job_scheduled_at` over it or even just keeping `run_at`. But between those two I'd prefer the former.

machisuji commented on docs/system-admin-guide/integrations/nextcloud/README.md:
```diff
@@ -348,7 +348,7 @@ You have setup the *Project folder* in both environments (Nextcloud and OpenProj
-   The result should show lines containing `bundle exec rake jobs:work`
+   The result should show lines containing `bundle exec bundle exec good_job start`
 
    b. If you don't have root access to the OpenProject server then you can check the following URL in your browser: `https://<your-openproject-server>/health_checks/all` (please insert the domain name of your OpenProject server). If your background workers are running, you should see a line like that `delayed_jobs_never_ran: PASSED All previous jobs have completed within the past 5 minutes`
```
There is still a mention of `delayed_job` here. Perhaps we should just rename this check to 'jobs_never_ran'.

machisuji commented on config/application.rb:
```diff
@@ -212,7 +212,18 @@ class Application < Rails::Application
-    config.active_job.queue_adapter = :delayed_job
+    config.active_job.queue_adapter = :good_job
+
+    config.good_job.preserve_job_records = true
```
We may want to configure the retention period too and perhaps use a lower one than the default because we have a whole lot of jobs. Perhaps it would make sense to make this configurable via `OpenProject::Configuration` too. Something like the following.

```
config.good_job.cleanup_preserved_jobs_before_seconds_ago = OpenProject::Configuration.good_job_record_retention_days.days
```

machisuji commented on config/application.rb:
```diff
@@ -212,7 +212,18 @@ class Application < Rails::Application
+    config.good_job.retry_on_unhandled_error = false
+    # config.good_job.on_thread_error = -> (exception) { Rails.error.report(exception) }
+    config.good_job.execution_mode = :external
+    config.good_job.queues = '*'
```
Just double checking, this is overriden by the `GOOD_JOB_QUEUES` env var, correct?

ba1ash replied:
No, it seems like this one has higher precedence. I will make it part of OpenProject configuration.

ba1ash replied:
No, it seems like this one has higher precedence than env var. I will make it part of the configuration.

machisuji commented on config/application.rb:
```diff
@@ -212,7 +212,18 @@ class Application < Rails::Application
+    # config.good_job.on_thread_error = -> (exception) { Rails.error.report(exception) }
+    config.good_job.execution_mode = :external
+    config.good_job.queues = '*'
+    config.good_job.max_threads = 20
```
I think it would be useful to make this configurable via OpenProject configuration too.

machisuji commented on config/initializers/cronjobs.rb:
```diff
@@ -1,15 +1,71 @@
+    {
+      'Cron::ClearOldSessionsJob': {
+        cron: '15 1 * * *', # runs at 1:15 nightly
+        class: 'Cron::ClearOldSessionsJob'
```
Can we use `Cron::ClearOldSessionsJob.name` here (and actually use the literal + .name in all the other entries too) so we can make sure we don't miss anything if jobs are renamed? Or would this throw an error elsewhere when the class parameter is invalid?

machisuji commented on db/migrate/20221026132134_bigint_primary_and_foreign_keys.rb:
```diff
@@ -60,7 +60,6 @@ class BigintPrimaryAndForeignKeys < ActiveRecord::Migration[7.0]
     CustomStyle => [:id],
     CustomValue => %i[id customized_id custom_field_id],
     Journal::CustomizableJournal => %i[id journal_id custom_field_id],
-    Delayed::Job => [:id],
```
I feel this migration will break things in the future, using literals which may or may not exist anymore in the future.
This is of course unrelated to your changes.
And while this may not be the case for delayed job in particular, this might perhaps break later migrations if we just remove entries here in the future. So I don't want to set a precedent.

I suggest keeping the delayed job entry around so we don't have to worry about any other consequences down the line and to set a good example for future changes touching this file. Looking at the code further below it should already work to change the line to:

```
'delayed_jobs' -> [:id], # delayed job removed in favour of good_job see WP #42 or PR #42
```

oliverguenther replied:
> This is not resolved, is it? The code is still removed. But of course Delayed::Job won't exist anymore. So this should just be done in plain SQL.
> execute_sql "DELETE FROM delayed_jobs WHERE handler LIKE '%job_class: Notifications::CleanupJob%'"

While technically correct, we don't need to remove delayed jobs as the table should be removed. @ba1ash I don't think this is happening currently correct?

ba1ash replied:
True, it is not happening right now. I will add it in this PR.

machisuji replied:
@ba1ash I still think this line should be preserved in the given form for the aforementioned reasons.

```
'delayed_jobs' -> [:id], # delayed job removed in favour of good_job see WP #42 or PR #42
```

machisuji commented on modules/github_integration/lib/open_project/github_integration/engine.rb:
```diff
@@ -85,9 +85,13 @@ class Engine < ::Rails::Engine
+      {
+        ClearOldPullRequestsJob: {
+          cron: '25 1 * * *', # runs at 1:25 nightly
+          class: 'ClearOldPullRequestsJob'
```
Same question here if we could use the class literal to make sure we notice if the class is renamed. So `class: ClearOldPullRequestsJob.name`.

ba1ash replied:
Yes, that makes perfect sense. Will do.

machisuji commented on modules/job_status/lib/open_project/job_status/engine.rb:
```diff
@@ -46,10 +46,16 @@ class Engine < ::Rails::Engine
+      {
+        'JobStatus::Cron::ClearOldJobStatusJob': {
+          cron: '15 4 * * *', # runs at 4:15 nightly
+          class: '::JobStatus::Cron::ClearOldJobStatusJob'
```
Again, should we use the literal here?

machisuji commented on modules/ldap_groups/lib/open_project/ldap_groups/engine.rb:
```diff
@@ -19,7 +19,14 @@ class Engine < ::Rails::Engine
+      {
+        'Ldap::SynchronizationJob': {
+          cron: '30 23 * * *', # Run once per night at 11:30pm
+          class: 'Ldap::SynchronizationJob'
```
Literal question again.

machisuji commented on modules/storages/app/workers/storages/manage_nextcloud_integration_job.rb:
```diff
@@ -0,0 +1,104 @@
+        .automatic_management_enabled
+        .includes(:oauth_client)
+        .find_each do |storage|
+          result = service_for(storage).call(storage)
```
It may be nit-picky but this formatting is offensive to my eyes. Could we perhaps just put either the whole block or just the storages in a method?

```
def storages
  ::Storages::Storage
    .automatic_management_enabled
    .includes(:oauth_client)
end

def perform
  storages.find_each do |storage|
    service_for(storage).call(storage).match(
      on_success: ->(_) { storage.mark_as_healthy },
      on_failure: ->(errors) { storage.mark_as_unhealthy(reason: errors.to_s) }
    )
  end
end
```

It's probably best to put the block inside the find_each into a method, too, though.

machisuji commented on modules/storages/db/migrate/20231009135807_remove_renamed_cronjobs.rb:
```diff
@@ -1,36 +0,0 @@
-
-class RemoveRenamedCronjobs < ActiveRecord::Migration[7.0]
-  def up
-    Delayed::Job.where("handler LIKE '%job_class: CleanupUncontaineredFileLinksJob%'").delete_all
```
This migration should be kept around just the same as the one RemoveNotificationCleanupJob. Again we should just use plain SQL to remove the jobs.

machisuji replied:
@ba1ash As mentioned above, this should be kept, not deleted.

machisuji commented on modules/storages/lib/open_project/storages/engine.rb:
```diff
@@ -142,14 +142,14 @@ def self.permissions
       menu :admin_menu,
            :storages_admin_settings,
-           { controller: '/storages/admin/storages', action: :index },
+           { controller: "/storages/admin/storages", action: :index },
```
This is probably your editor changing this manually. And the question whether to use single or double quotes can lead to heated discussions. But I would find it best if these just weren't changed at all in all lines that haven't changed semantically. Is there a way to prevent your editor from doing this?

ba1ash replied:
It is not my editor. Usually I apply `rubocop -a` to fix style before pushing. That is the outcome of it. I see your point. The PR size is huge already and it adds more to it. My bad. In the future I'll try to keep separate and not include lines that semantically has not changed.

machisuji commented on modules/storages/lib/open_project/storages/engine.rb:
```diff
@@ -273,21 +273,28 @@ def self.permissions
+      {
+        'Storages::CleanupUncontaineredFileLinksJob': {
+          cron: "06 22 * * *",
+          class: "Storages::CleanupUncontaineredFileLinksJob"
```
literal?

machisuji commented on modules/storages/lib/open_project/storages/engine.rb:
```diff
@@ -273,21 +273,28 @@ def self.permissions
+
+        'Storages::ManageNextcloudIntegrationJob': {
+          cron: "*/5 * * * *",
+          class: "Storages::ManageNextcloudIntegrationJob"
```
literal?

machisuji commented on modules/storages/spec/workers/storages/cleanup_uncontainered_file_links_job_spec.rb:
```diff
@@ -32,39 +32,37 @@
-    allow(OpenProject::Configuration)
-      .to receive(:attachments_grace_period)
-            .and_return(grace_period)
+  describe '#perfrom' do
```
typo, it should be `'#perform'`

machisuji commented on modules/webhooks/lib/open_project/webhooks/engine.rb:
```diff
@@ -51,6 +51,13 @@ class Engine < ::Rails::Engine
+      {
+        CleanupWebhookLogsJob: {
+          cron: '*/1 * * * *', # runs at 5:28 on Sunday
+          class: 'CleanupWebhookLogsJob'
```
literal? We seem to be using it for the cron key already as well here.

machisuji commented on spec/features/notifications/reminder_mail_spec.rb:
```diff
@@ -9,10 +9,10 @@
   let(:watched_work_package) { create(:work_package, project:, watcher_users: [current_user]) }
   let(:involved_work_package) { create(:work_package, project:, assigned_to: current_user) }
-  # The run_at time of the delayed job used for scheduling the reminder mails
+  # GoodJob::Job#scheduled_at is used for scheduling the reminder mails
```
This goes back to the comment further up. Perhaps we could just call the method `job_scheduled_at` in `ApplicationJob` and reference that here?

machisuji commented on spec/workers/work_packages/apply_working_days_change_job_spec.rb:
```diff
@@ -1811,26 +1811,4 @@
     end
   end
-
-  describe '.scheduled?' do
```
Related to [this](https://github.com/opf/openproject/pull/12998/files#diff-254057ccc73a0d0229fc0f8cb7c67b1f26c2539c43e5ad90a59f3eec0e5dc7f3R46) perhaps it makes sense to still have a (although universal) `.scheduled?` kind of method in application job and use that in the linked spot and test it in a separate spec.

ba1ash replied:
I would not create it for now and try to use GoodJob concurrency control mechanism to control the number of jobs in the que. The latter should be more reliable.

machisuji reviewed (changes requested):
Thanks for all the work, @ba1ash! GoodJob looks really promising. I do have some remarks, suggestions and questions. Could you please have a look?

oliverguenther committed:
Fix testing value for cron

ba1ash committed:
Merge branch 'dev' into task/48717-replace-delayedjob

ba1ash committed:
Use plain sql to cleanup removed cron_jobs.

ba1ash committed:
React on comments from review.

- Do not use string literals for job class names. Use `class.name` instead.
- Rename `ApplicationJob#good_job_scheduled_at` to `ApplicationJob#job_scheduled_at` to be backend agnostic.
- update queries in bin/check-worker-liveness to use good_jobs table
- Set good_job config options through appropriate OpenProject::Configuration
- Remove delayed_jobs table.
- Update health_check docs.

ba1ash committed:
Merge remote-tracking branch 'origin/dev' into task/48717-replace-delayedjob

ba1ash committed:
Make execute_sql work.

ba1ash committed:
Fix #upsert_status race condition.

Rescue ActiveRecord::RecordNotUnique and then retry.

ba1ash committed:
Merge branch 'dev' into task/48717-replace-delayedjob

ba1ash committed:
Remove TODO comment in favor of dedicated ticket.

https://community.openproject.org/wp/53122

machisuji commented on modules/job_status/app/workers/job_status/application_job_with_status.rb:
```diff
@@ -77,7 +74,16 @@ def upsert_status(status:, **args)
         resource.attributes = build_status_attributes(args.merge(status:))
       end
 
+      # There is a possible race condition because of unique delayed_job_statuses.job_id
```
You will hate me, but this reminds me that it would be prudent to rename the `delayed_job_statuses` table to simply `job_statuses` and update the table_name in `modules/job_status/app/models/job_status/status.rb`. And consequently the reference here should be updated too.

ba1ash replied:
The renaming absolutely makes sense. I will do it.

machisuji reviewed (changes requested):
Thanks for addressing most of the changes I requested. Could you please have a look at the few new comments I made?

Also one question: What happens to existing delayed jobs in the course of this migration? As far as I can tell they would just be deleted and not translated into new good jobs.

But wouldn't that be necessary? I'm just imagining a scenario where a bunch of `Journals::CompletedJob`s are scheduled `Setting.journal_aggregation_time_minutes` later while the update is being performed.

I just want to make sure it wouldn't be a problem if these jobs are lost. Would it it just be notifications (including web hooks) that are lost? @ulferts what do you say? Is that ok or shouldn't we migrate existing delayed jobs into the new good_job table? One could argue that even lost webhooks aren't optimal. But I'm wondering if I'm missing anything else.

ba1ash committed:
Restore delayed_job related migration statements.

ba1ash committed:
Rename delayed_job_statuses to job_statuses.

ba1ash committed:
Use execute over execute_sql.

Due to no need in sanitization.

ba1ash commented:
> Thanks for addressing most of the changes I requested. Could you please have a look at the few new comments I made?
> 
> Also one question: What happens to existing delayed jobs in the course of this migration? As far as I can tell they would just be deleted and not translated into new good jobs.
> 
> But wouldn't that be necessary? I'm just imagining a scenario where a bunch of `Journals::CompletedJob`s are scheduled `Setting.journal_aggregation_time_minutes` later while the update is being performed.
> 
> I just want to make sure it wouldn't be a problem if these jobs are lost. Would it it just be notifications (including web hooks) that are lost? @ulferts what do you say? Is that ok or shouldn't we migrate existing delayed jobs into the new good_job table? One could argue that even lost webhooks aren't optimal. But I'm wondering if I'm missing anything else.

True. They will be deleted.
My assumption was that we can afford losing jobs. If we can't I will prepare a migration.

ba1ash committed:
Merge branch 'dev' into task/48717-replace-delayedjob

ba1ash committed:
Migrate 'in queue' jobs to good_jobs table.

Update good_job version as well.

ba1ash committed:
Remove delayed_job dependecy. Lock jobs.

- Remove delayed_job dependecy completely from Gemfile.
- Lock jobs during migration from delayed_jobs to good_jobs using SELECT FOR UPDATE.
  It means that potentially not shut down DJ worker will not be able to
  pick them up for performing during the migration process. After
  migration has been finished the jobs in question will be in
  good_jobs table and delayed_jobs table will be removed. So,
  duplicate execution should not happen.

ba1ash committed:
Merge branch 'dev' into task/48717-replace-delayedjob

machisuji reviewed (approved):
Awesome, thanks for all the effort! 👍
